pubsublite/internal/wire/subscriber.go

```diff
@@ -399,3 +401,107 @@ func newMultiPartitionSubscriber(subFactory *singlePartitionSubscriberFactory) *
+// listen to its assigned partition numbers and dynamically add/remove
+// singlePartitionSubscribers.
+type assigningSubscriber struct {
+	// Immutable after creation.
+	subFactory *singlePartitionSubscriberFactory
+	assigner   *assigner
+
+	// Fields below must be guarded with mu.
+	// Subscribers keyed by partition number. Updated as assignments change.
+	subscribers map[int]*singlePartitionSubscriber
+
+	compositeService
+}
+
+func newAssigningSubscriber(assignmentClient *vkit.PartitionAssignmentClient, genUUID generateUUIDFunc, subFactory *singlePartitionSubscriberFactory) (*assigningSubscriber, error) {
+	as := &assigningSubscriber{
+		subFactory:  subFactory,
+		subscribers: make(map[int]*singlePartitionSubscriber),
+	}
+	as.init()
+
+	assigner, err := newAssigner(subFactory.ctx, assignmentClient, genUUID, subFactory.settings, subFactory.subscriptionPath, as.handleAssignment)
+	if err != nil {
+		return nil, err
+	}
+	as.assigner = assigner
+	as.unsafeAddServices(assigner)
+	return as, nil
+}
+
+func (as *assigningSubscriber) handleAssignment(partitions partitionSet) error {
+	as.mu.Lock()
+	defer as.mu.Unlock()
+
+	// Handle new partitions.
+	for _, partition := range partitions.Ints() {
+		if _, exists := as.subscribers[partition]; !exists {
+			subscriber := as.subFactory.New(partition)
+			if err := as.unsafeAddServices(subscriber); err != nil {
+				// Occurs when the assigningSubscriber is stopping/stopped.
+				return err
+			}
+			as.subscribers[partition] = subscriber
+		}
+	}
+
+	// Handle removed partitions.
+	for partition, subscriber := range as.subscribers {
+		if !partitions.Contains(partition) {
+			as.unsafeRemoveService(subscriber)
+			// Safe to delete map entry during range loop:
+			// https://golang.org/ref/spec#For_statements
+			delete(as.subscribers, partition)
+		}
+	}
+	return nil
+}
+
+// Subscriber is the client interface exported from this package for receiving
+// messages.
+type Subscriber interface {
+	Start()
+	WaitStarted() error
+	Stop()
+	WaitStopped() error
+}
+
+// NewSubscriber creates a new client for receiving messages.
+func NewSubscriber(ctx context.Context, settings ReceiveSettings, receiver MessageReceiverFunc, region, subscriptionPath string, opts ...option.ClientOption) (Subscriber, error) {
+	if err := ValidateRegion(region); err != nil {
+		return nil, err
+	}
+	if err := validateReceiveSettings(settings); err != nil {
+		return nil, err
+	}
+	subClient, err := newSubscriberClient(ctx, region, opts...)
+	if err != nil {
+		return nil, err
+	}
+	cursorClient, err := newCursorClient(ctx, region, opts...)
+	if err != nil {
+		return nil, err
+	}
+
+	subFactory := &singlePartitionSubscriberFactory{
+		ctx:              ctx,
+		subClient:        subClient,
+		cursorClient:     cursorClient,
+		settings:         settings,
+		subscriptionPath: subscriptionPath,
+		receiver:         receiver,
+	}
+
+	if len(settings.Partitions) > 0 {
+		return newMultiPartitionSubscriber(subFactory), nil
```

Comment:
I would actually delete this class entirely. assigningSubscriber is a generalization of this, where the "assigner" only ever produces a single set of values.

Reply:
True.  Although we can avoid extra complexity in the `assigner` to handle this mode, avoid creating the PartitionAssignmentClient, etc. So I might leave `newMultiPartitionSubscriber`, as it's only a few lines of extra code (+tests) anyway.